File: src/js/4_BehavioralPatterns/11_Visitor/visitor_exercise.js
class Integer
{
  constructor(value)
  {
    this.value = value;
  }

  accept(visitor)
  {
    visitor.visitValue(this);
  }
}

class BinaryExpression
{
  constructor(lhs, rhs)
  {
    this.lhs = lhs;
    this.rhs = rhs;
  }
}

class AdditionExpression extends BinaryExpression
{
  constructor(lhs, rhs) {
    super(lhs, rhs);
  }

  accept(visitor)
  {
    visitor.visitAddition(this);
  }
}

class MultiplicationExpression extends BinaryExpression
{
  constructor(lhs, rhs) {
    super(lhs, rhs);
  }

  accept(visitor)
  {
    visitor.visitMultiplication(this);
  }
}

class ExpressionPrinter
{
  constructor()
  {
    this._buffer = [];
  }

  visitValue(value)
  {
    this._buffer.push(value.value);
  }

  visitAddition(ae)
  {
    this._buffer.push('(');
    ae.lhs.accept(this);
    this._buffer.push('+');
    ae.rhs.accept(this);
    this._buffer.push(')');
  }

  visitMultiplication(me)
  {
    me.lhs.accept(this);
    this._buffer.push('*');
    me.rhs.accept(this);
  }

  toString()
  {
    return this._buffer.join('');
  }
}

describe('visitor', function()
{
  it('simple addition', function()
  {
    let simple = new AdditionExpression(
      new Integer(2), new Integer(3)
    );
    let ep = new ExpressionPrinter();
    ep.visitAddition(simple);
    expect(ep.toString()).toEqual('(2+3)');
  });

  it('product of addition and value', function()
  {
    const expr = new MultiplicationExpression(
      new AdditionExpression(new Integer(2), new Integer(3)),
      new Integer(4)
    );
    let ep = new ExpressionPrinter();
    ep.visitMultiplication(expr);
    expect(ep.toString()).toEqual('(2+3)*4');
  });
});

File: src/java/4_BehavioralPatterns/12_Visitor/VisitorExercise.java
package com.activemesa.behavioral.visitor.exercise;

abstract class ExpressionVisitor
{
  abstract void visit(Value value);
  abstract void visit(AdditionExpression ae);
  abstract void visit(MultiplicationExpression me);
}

abstract class Expression
{
  abstract void accept(ExpressionVisitor ev);
}

class Value extends Expression
{
  public int value;

  public Value(int value)
  {
    this.value = value;
  }

  @Override
  void accept(ExpressionVisitor ev)
  {
    ev.visit(this);
  }
}

class AdditionExpression extends Expression
{
  public Expression lhs, rhs;

  public AdditionExpression(Expression lhs, Expression rhs)
  {
    this.lhs = lhs;
    this.rhs = rhs;
  }

  @Override
  void accept(ExpressionVisitor ev)
  {
    ev.visit(this);
  }
}

class MultiplicationExpression extends Expression
{
  public Expression lhs, rhs;

  public MultiplicationExpression(Expression lhs, Expression rhs)
  {
    this.lhs = lhs;
    this.rhs = rhs;
  }

  @Override
  void accept(ExpressionVisitor ev)
  {
    ev.visit(this);
  }
}

class ExpressionPrinter extends ExpressionVisitor
{
  private StringBuilder sb = new StringBuilder();

  @Override
  void visit(Value value)
  {
    sb.append(value.value);
  }
  @Override
  void visit(AdditionExpression ae)
  {
    sb.append("(");
    ae.lhs.accept(this);
    sb.append("+");
    ae.rhs.accept(this);
    sb.append(")");
  }
  @Override
  void visit(MultiplicationExpression me)
  {
    me.lhs.accept(this);
    sb.append("*");
    me.rhs.accept(this);
  }

  @Override
  public String toString()
  {
    return sb.toString();
  }
}